### ISSUE TITLE
extract polygon vertices: total number of points incorrect

### ISSUE THREAD

**Issue by pigreco** [mbloch/mapshaper#307]
Hello,
I use this script [0]:
```
time node --max-old-space-size = 4192 `which mapshaper` encoding = utf-8 Com01012018_g_wgs84.shp -points vertices -explode -o vertici_g_mapsh.shp
```

to extract the vertices of a multipolygon shapefile [1],
the number of vertices (844.696) does not match the number obtained with postgis using the query:

`SELECT SUM (ST_NPOINTS (GEOM)) FROM TABLE_NAME;`  → 853.700 points

The difference is 9004 points, the duplicate points are about 1/2 of the total so it is not a topological problem.

thank you

[0] mapshaper:
https://github.com/pigreco/benchmark/blob/master/img/estrai_v/mapshaper_01.png
[1] shapefile:
https://github.com/pigreco/benchmark/blob/master/dati/com01012018_g_WGS84_ok.zip

**Comment by mbloch** [comment]
The `-points vertices` command gives you the unique point locations within each feature. There will be duplication between adjacent features (because of topology), but not within a feature.

The postgis query doesn't remove duplication of any kind. So the first and last vertices of polygon rings are both retained, even though they have the same coordinates. Also, duplicate points within a ring, or between rings in the same feature, are also retained by postgis. Mapshaper's command removes duplicates in all of these cases.

Perhaps mapshaper should also allow conversion without any deduplication... would that be useful to you?

**Comment by aborruso** [comment]
> Perhaps mapshaper should also support conversion without any deduplication... would that be useful to you?

It could be useful maybe to have always the primitives to come back to the polygons

**Comment by mbloch** [comment]
When converting from polygons or multi-polygons to points, you lose the ring structure (points are no longer grouped by rings), so the round-trip conversion is a bit complicated.

If there's an application related to map making, I will try to support it...  I'm less interested in making mapshaper into a general-purpose GIS toolkit.

**Comment by pigreco** [comment]
>Perhaps mapshaper should also support conversion without any deduplication... would that be useful to you?

yes, it would be useful if two points had the same x and y but z different

thank you

**Comment by mbloch** [comment]
Currently, mapshaper does not support Z data. "M" and "Z" data is ignored when Shapefiles containing those data fields are imported. So adding support for Z data would be a precondition for your use case, @pigreco.

**Comment by mbloch** [comment]
In commit 666d99ffc7bb7d7bf27cb51fbde5  I added `-points vertices2`, which doesn't remove duplicate coordinates within a feature. Adding support for M and Z data is a low priority for me, so I'm closing this issue.

**Comment by mbloch** [comment]
I just published v0.4.103, which includes `-points vertices2`.

**Comment by aborruso** [comment]
thank you @mbloch !
